### PR TITLE
Image Upload: make an image request cancellable

### PR DESCRIPTION
**Pull request by jaclync**
"Allow an image request to be cancelled when used in a reusable cell" for #1713 

## Changes

- Before this PR, `ImageDownloader` is a class of `Kingfisher` that downloads an image. This PR created an `ImageDownloader` protocol and made `Kingfisher.ImageDownloader` conform to the protocol
- In `ImageService` protocol, returned a `ImageDownloadTask` for an image download request
  - `ImageDownloadTask` conforms to a `Cancellable` protocol (built-in iOS 13+ but we can't use it yet)
- In `ProductUIImageLoader` protocol, returned a `Cancellable` for a UI image request similar to `ImageService`
  - In its implementation `DefaultProductUIImageLoader`, kept track of the active image requests and cancelled them on `deinit`
- In `ProductImageCollectionViewCell`, added a property of `Cancellable` and cancelled the optional task on `prepareForReuse`
  - Updated its usage in `ProductImagesCollectionViewController` and `ProductImagesCollectionViewDataSource`
- In `AztecTextViewAttachmentHandler`, kept track of the active image requests and cancelled them on `deinit`
- In WP media library data source (`WordPressMediaLibraryPickerDataSource`), created a new class `CancellableMedia` that includes a `Media` and `Cancellable` so that a media can be cancelled to replace the media items used in `WordPressMediaLibraryPickerDataSource`
  - Reverted `Media` back to a struct!
- Updated unit tests

## Testing

It's a bit harder to test this change, but it's possible in WP media library picker with `Wormholy`:

- Please make sure the image cache for the WP media library is empty, like reinstalling the app or switching to a different store that you haven't opened the WP media library before.
- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- (Please check the bold text before tapping the option) Tap "WordPress Media Library", then **make sure you've seen some cells whose image haven't been loaded and quickly leave the screen**
- Leave edit product screen, and go to "My store" tab
- Tap settings > "Launch Wormholy debug" and filter by "uploads" for image requests --> there should be some requests that never complete (cancelled) like in the screenshot below (you can try the same steps where you leave the picker before some cell images are loaded on `develop`, and all image requests complete)
 
`develop` | PR branch
-- | --
Simulator Screen Shot - iPhone 11 - 2020-04-09 at 14 17 19 | Simulator Screen Shot - iPhone 11 - 2020-04-09 at 14 22 25

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Tools/ImageService/ImageDownloader.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import UIKit
+
+/// Used for any activity or action that may be canceled.
+/// `Cancellable` is also available starting iOS 13, please delete when the project is iOS 13+.
```
Comment (pmusolino) [inline WooCommerce/Classes/Tools/ImageService/ImageDownloader.swift:4]: Maybe adding a `TODO` keyword can help us in the future to find it easily

File: WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift [hunk 2]
```diff
@@ -2,37 +2,39 @@ import MobileCoreServices
...
         let imageService = ServiceLocator.imageService
-        imageService.retrieveImageFromCache(with: url) { (image) in
+        imageService.retrieveImageFromCache(with: url) { [weak self] (image) in
             if let image = image {
                 completionHandler(image, nil)
                 return
             }
 
-            imageService.downloadImage(with: url, shouldCacheImage: true) { (image, error) in
+            self?.cancellableTask = imageService.downloadImage(with: url, shouldCacheImage: true) { (image, error) in
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift:21]: I believe this should be:
```suggestion
            self?.cancellableTask = imageService.downloadImage(with: url, shouldCacheImage: true) { [weak self] (image, error) in
```

File: WooCommerce/Classes/ViewRelated/Products/Media/WordPressMediaLibraryPickerDataSource.swift [hunk 3]
```diff
@@ -22,8 +22,7 @@ extension WordPressMediaLibraryPickerDataSource {
         }
 
         func cancelImageRequest(_ requestID: WPMediaRequestID) {
-            // TODO-2073: implement image request cancelling once the image fetching is not tied to `Media` since it needs to keep
-            // track of the task.
+            // No image is shown for the media group.
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Media/WordPressMediaLibraryPickerDataSource.swift:25]: Is this still part of https://github.com/woocommerce/woocommerce-ios/issues/2073?

File: WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift:21]: hmm I don't see the use of `self` in the completion block, and `completionHandler` in the completion block isn't mutated? lemme know if I missed anything!

File: WooCommerce/Classes/ViewRelated/Products/Media/WordPressMediaLibraryPickerDataSource.swift (hunk 3 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Media/WordPressMediaLibraryPickerDataSource.swift:25]: great catch! I removed this TODO item from https://github.com/woocommerce/woocommerce-ios/issues/2073

**Comment by jaclync** [comment]
@pmusolino I updated the PR, merging now - I'm happy to update based on your followup answer https://github.com/woocommerce/woocommerce-ios/pull/2114#discussion_r407487612 separately later!

File: WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift (hunk 2 above)
Comment (pmusolino) [reply WooCommerce/Classes/ViewRelated/Products/Media/Media+WPMediaAsset.swift:21]: You are right, `self` is not used within the block. Please ignore what I said 😄

**Pull request merged**